packages/lazy-import/lib/index.js

```diff
@@ -0,0 +1,129 @@
+ *
+ * @param {string} arg A string that you might pass to `npm install`.
+ *
+ * @return {string} Module to use as local package alias.
+ */
+function getLocalModuleName( arg ) {
+	return '@wordpress/lazy-import.' + md5( arg );
+}
+
+/**
+ * Installs npm package by arg name at the given prefix path. Creates the prefix
+ * path if it doesn't yet exist.
+ *
+ * @param {string} arg   Package install arg (name and version specification).
+ * @param {string} alias Alias to use for local module name.
+ *
+ * @return {Promise<void>} Promise resolving once package is installed.
+ */
+async function install( arg, alias ) {
+	await execa( 'npm', [ 'install', '--no-save', alias + '@npm:' + arg ] );
+}
+
+/**
+ * Given an arg string as would be passed to `npm install`, requires the package
+ * corresponding to the name and specifier parsed from the arg. If the package
+ * is not installed (or not installed at that version), the package is installed
+ * to a temporary directory at `node_modules/.wp-lazy` relative to where the
+ * current working directory.
+ *
+ * @param {string}                       arg       A string that you might pass
+ *                                                 to `npm install`.
+ * @param {Partial<WPLazyImportOptions>} [options] Optional options object.
+ *
+ * @return {Promise<NodeRequire>} Promise resolving to required module.
+ */
+async function lazyImport( arg, options = {} ) {
+	const { rawSpec, name } = npmPackageArg( arg );
+
+	if ( ! name ) {
+		throw new TypeError(
+			`Unable to parse package name from \`${ arg }\`.`
+		);
+	}
+
+	const localModule = getLocalModuleName( arg );
+
+	// Try first from the temporary install path, since the second attempt will
+	// need to verify both availability and version. Version isn't necessary to
+	// account for in this first attempt.
+	try {
+		return require( localModule );
+	} catch ( error ) {
+		if ( error.code !== 'MODULE_NOT_FOUND' ) {
+			throw error;
+		}
+	}
+
+	try {
+		const resolved = require( name );
+
+		const { version } = require( join( name, 'package.json' ) );
+		if ( semver.satisfies( version, rawSpec ) ) {
+			// Only return with the resolved module if the version is valid per
+			// the parsed arg. Otherwise, fall through to install stage.
+			return resolved;
+		}
+	} catch ( error ) {
+		if ( error.code !== 'MODULE_NOT_FOUND' ) {
+			throw error;
+		}
+	}
+
+	// If this point is reached, the module cannot be found and must be
+	// installed.
+	if ( options.onInstall ) {
+		options.onInstall();
+	}
+
+	try {
+		await install( arg, localModule );
+	} catch ( error ) {
+		// Format connectivity error to one which can be easily evaluated.
+		if (
+			error.stderr &&
+			error.stderr.startsWith( 'npm ERR! code ENOTFOUND' )
+		) {
+			error = new Error( 'Unable to connect to NPM registry' );
+			error.code = 'ENOTFOUND';
+		}
+
+		throw error;
+	}
+
+	// Delete cache from prior `require` attempt.
+	delete require.cache[ require.resolve( localModule ) ];
```

Comment:
In my testing, this line doesn't work when something goes wrong. We can remove it. If the package is properly installed then `require` in the next line works regardless.

Reply:
I was puzzled about this, since it seems like given the [intention of `require.cache` to bypass a second lookup](https://nodejs.org/api/modules.html#modules_caching) would conflict with the intention of the following line to try the import once more.

But I guess it depends when a module is cached. Notably, the use-case here can assume that the previous attempt would have been a failure.

Briefly testing this in the terminal, it does appear that a module will not be cached if it couldn't be resolved:

```
⇒ node -e "try { require( './abc.js' ); } catch {} console.log( require.cache );"
[Object: null prototype] {}
```

So, as you mention, I suppose it should be safe to remove.